pkg/webhook/validator.go

```diff
@@ -149,10 +149,13 @@ func (v *Validator) ValidateCronJob(ctx context.Context, c *duckv1.CronJob) *api
 		ImagePullSecrets:   imagePullSecrets,
 	}
 
-	return v.validatePodSpec(ctx, ns, c.Kind, c.APIVersion, c.ObjectMeta.Labels, &c.Spec.JobTemplate.Spec.Template.Spec, opt).ViaField("spec.jobTemplate.spec.template.spec")
+	return v.validatePodSpec(ctx, ns, c.Kind, c.APIVersion, c.ObjectMeta.Labels, &c.Spec.JobTemplate.Spec.Template.Spec, opt, false).ViaField("spec.jobTemplate.spec.template.spec")
 }
 
-func (v *Validator) validatePodSpec(ctx context.Context, namespace, kind, apiVersion string, labels map[string]string, ps *corev1.PodSpec, opt k8schain.Options) (errs *apis.FieldError) {
+func (v *Validator) validatePodSpec(ctx context.Context, namespace, kind, apiVersion string, labels map[string]string, ps *corev1.PodSpec, opt k8schain.Options, isPod bool) (errs *apis.FieldError) {
```

Comment:
This is serving a very specific use case, I thought you wanted to disable the mutation of the digest for all the resources.

Reply:
I was thinking about it and realized that this is not generic enough in its form.

What I think this should be is making digest behavior configurable to specify when the image is replaced with digest - in all objects (Deployment, StatefulSet, CronJob, Job, ... as well as Pod), pod or never.

So from configuration perspective it could be the following instead of current setting in this PR:

```
update-image-to-digests: always
update-image-to-digests: pod
update-image-to-digests: never
```
(defaulting to `always`)

This way it's possible to configure whether top-level object like Deployment gets mutated (default, fastest, secure), only the Pods get mutated (which is slower, but should still be as secure), no objects get mutated (which is less secure).

I think for the setup that we have we would like the Pod-level mutation to happen, but others may want something else, so having 3 options should be most configurable.

As for passing this line specifically and `isPod`, I'm simply passing if this was called for Pod or another type of object so the code can determine its behavior.

Would that make more sense?

Reply:
I'm not the k8s expert but "never" sure sounds risky.

Reply:
I have changed this approach and pushed the change. The only difference from my comment above is I called th setting `update-image-to-digest`. It acceps settings `always`, `pod` and `never`.

This way it defaults to current behavior, but allows configuring whether to patch Pods only or not patch at all.

I think this is a reasonable feature and not just specific to our use case. In many cases it's fine to patch derived objects such as Pod, but it is more complex to have objects managed by GitOps patched by other controllers / operators.

I find the pattern of patching Pods instead of higher-level objects like Deployment/StatefulSet/CronJob/Job quite common. In this case it seems like a reasonable compromise to me - patching Pods prevents image tag updates between validation and pod starting from being a security issue.

Reply:
I'm really curious why you would want to patch Pod A of Deployment X to be different than say Pod B. Just trying to understand why this is common. Naively this seems to me as an antipattern, so just trying to understand :)

Reply:
@imjasonh Do you have any ideas how to handle this use case ? We believe you have more thoughts on why we should  avoid using tags, and instead stick to digests.

Reply:
I've been looking more into this, trying to understand how this is a problem for me, how other tools to it and think of ways forward. It helped me clear my mind.

The projects I work on use GitOps and ArgoCD to deploy everything - there's a repo that has everything included, dependencies vendored in and rendered as YAML files. ArgoCD then deploys those files and makes sure that all Kubernetes objects in manifests that are in the Git repo (such as Deployments, StatefulSets etc). Some of our deployments are upstream projects we use, either via helm charts or in other ways.

When `policy-controller` and its mutating webhooks, whenever ArgoCD updates or creates objects like Deployments, it patches the image, and this is the live manifest of a Deployment. ArgoCD sees this as a difference between desired and live manifest - it expects `some/image:v1.23` while Kubernetes returns `docker.io/some/image@sha256:...`. It causes ArgoCD to be consistently trying to apply the difference and never considers the application to be synchronized and up to date.

This only happens for any objects that do not use digests and in an ideal world I would simply only use image digests. We have over 100 apps we are including in our environments, many of which are just upstream charts or manifests. I did a quick grep and am seeing several hundred different images not referenced via digests across all apps and environments.

In many cases there's no simple and standard way to set images to digests - so I'd rather avoid having to find ways to patch the YAML manifests to patch any image tags to image digests to use `policy-controller`.

Unfortunately this is often the case when trying to apply a new tool to existing environments.

As for the why - I think the problem is that ArgoCD is expecting manifests to be applied without mutations (at least the level of objects ArgoCD manages - i.e. it ignores any changes/mutations at the level of ReplicaSets/Pods), while `policy-controller` today applies mutations to all objects in namespaces it controls. I am not sure why I seem to be the first one reporting it - I would expect this to be a more common problem with GitOps approach. ArgoCD (and I suppose other CD tools) work best when mutations are stored in new fields, not replacing existing fields.

I see ArgoCD allows ignoring fields managed by other controllers, but I don't think changes from mutating webhooks are tracked as managed fields or if there is an easy way to do so. That could be too complex to implement, though and I am not sure if it would work in this specific case - docs are here: https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#application-level-configuration

So at that point I was curious about other options. I don't think we'd be willing to look into alternatives for ArgoCD just because of this issue. But I checked [customizing how ArgoCD is diffing objects](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/) in ArgoCD and if it could help, but this causes another problem - if the image changes are ignored and are the only change in a Deployment, ArgoCD will just not apply the new image.

It seems that [connaisseur](https://github.com/sse-secure-systems/connaisseur/) started by only patching and validating Pods. There's a feature called [Automatic Child Approval](https://github.com/sse-secure-systems/connaisseur/blob/5574d34f86dfba8622d173f5cc159bd64baa6038/docs/features/automatic_child_approval.md) that does something similar to policy-controller's current behavior - replacing image with digest at the parent level.

Ideally, I think the solution should be done at the node and/or Kubernetes level - ideally Kubernetes in managed clouds should provide a way to validate images (either using webhooks or similar mechanism) so `policy-controller` could simply reject images after they're pulled on the node, but before pod is started. I remember someone at SigstoreCon mentioned that this is how they are doing image validation, but it works because they are not using managed Kubernetes clusters. Unfortunately, this is not something that would be available in a reasonable timeframe, even if work on it would begin now.

Another solution would be for `policy-controller` to resolve images at Deployment level but store them as PodSpec's annotation or any other location and then use that at the Pod level. This annotation could also have a signature based on webhook's private/public key pair. I think this could be too complex for the problem I am trying to find a solution to.

So, this makes me go back to my solution. I think it's reasonable, it allows 3 modes of operation - current mode, which always updates all objects, pod only updates and no updates. The first one makes sense for new workloads. The `pod` setting makes it behave more like [connaisseur](https://github.com/sse-secure-systems/connaisseur/) that it only patches Pods and not parent objects, but it also validates parent objects, which is good as it makes applying a Deployment with invalid images fail right away. I am not that passionate about `never` option, but I added it for completeness.

Regarding pod level patching - the main problem there is additional round trips to OCI registries. Resolving tag to digest at each `Pod` level is going to add more load, but I think that can be improved by adding caching.

I already created #278 as `policy-controller` is resolving tag to digest and then checking digest's signatures for every key in key-based validation. I think there's some cache scaffolding in the code already. Perhaps caching retrieving signatures for a digest and resolving tag to digest could be cached and then the overhead of patching each pod instead of patching the parent would not be as high.

As for pods potentially having different patches applied - in most cases the tag-based images are pointing to specific releases, not rolling tags. So if the images do not change, each pod would have the same patch applied to it. This is also how creating deployments with tag-based images works today - if I deploy `ubuntu:latest`, a pod created one year ago and today would potentially have different image.